### PR TITLE
Don't put NUL in the buffer in VT mode

### LINKED ISSUE #1825: Bug Report: CentOS system ssh cursor behavior bug
# Environment

```none
Windows build number: Microsoft Windows [Version 10.0.18362.207]
Windows Terminal version (if applicable): Version: 0.2.1831.0

Any other software?
```

# Steps to reproduce

1. ssh to a centOS based system (it has to be a CentOS Linux)
2. open a Linux screen session
3. press backspace when there is nothing in the current command line

# Expected behavior

Nothing happens

# Actual behavior

The location of the changes. The display of the text is weird if you type something.
Similar problem link: https://github.com/microsoft/WSL/issues/1436

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

Filter NUL from the stream when emitted in VT processing mode. Linux terminals can't really have a NUL in their "buffer" - the NUL characters are just ignored. For this particular bug, `screen` was writing a long sequence of NUL bytes as effectively a timeout, so that it could "flash" the screen.

## PR Checklist
* [x] Closes #1825
* [x] I work here
* [x] Tests added/passed
* [n/a] Requires documentation to be updated

## Detailed Description of the Pull Request / Additional comments

This PR is here for discussion - I'm a tad bit worried that it might regress some scenario unintentionally. `cmd.exe` uses VT processing for its own scripts and internal commands, is there something here we might be breaking? I _doubt_ it, but I'm unsure.

## Validation Steps Performed
* tested on a centos VM with `screen` to ensure this actually fixes the problem at hand.

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
Pains me to say, but should this be in WriteChars...?

File: tools/bx.ps1 [hunk 1]
```diff
@@ -20,6 +20,11 @@ $targets = $Metaproj.Project.Target
 # For Conhost\Server, this will match:
 #   [Conhost\Server, Conhost\Server:Clean, Conhost\Server:Rebuild, Conhost\Server:Publish]
 $matchingTargets = $targets | Where-Object { $_.MSBuild.Condition -eq $msBuildCondition }
+if ($matchingTargets.length -eq 0)
```
Comment (DHowett-MSFT) [inline tools/bx.ps1]: hold up, are you gonna conflict yourself? why does every PR have changes to bx ☹️

Comment (zadjii-msft) [reply tools/bx.ps1]: AH that shouldn't be there!

The bcz change should be, but I was leaving the bx change in the PR it came from, which unfortunately, isn't merged, and I keep temporarily copying locally so I can build things

Comment (zadjii-msft) [reply tools/bx.ps1]: (thanks)

**Comment by zadjii-msft** [comment]
So I don't know how this could really break backcompat. In general, we're talking about cmd.exe. For other "legacy" applications, cmd.exe is going to revert to "normal"/non-VT mode before launching the child process, so they won't be affected by this change.

Apps that _are_ using VT mode probably aren't doing this anyways, and if they're using VT mode, then they probably expect this fixed behavior, not the broken behavior.

cmd.exe however was changed to be in VT mode itself, so this change will apply to cmd.exe. Things that it might affect:
1. processing `.bat`/`.cmd` files
2. processing commandlines
3. Pasting input with a NUL

At least for 2, we don't need to worry. While we can insert other C0 chars into the commandline, (e.g. <kbd>ctrl+[</kbd> -> '\x1b` in a cooked read), it seems like we can't use this same input mechanism to input a NUL. <kbd>ctrl+@</kbd> does nothing, and <kbd>ctrl+Space</kbd> just inserts a space. So nothing to worry about there.

For 3, currently when you paste input with a NUL in it, we treat that NUL as the end of the string. So you can't paste a NUL into the cooked input line. This is unchanged with my delta, so nothing to worry about here.

Then, as far as 1 goes, it seems like when cmd is processing a `.bat` file with a NUL in the middle of a line, it just ignores the rest of the line. So for ex, this hacked together bat file executes like the following:
image
image

This behavior looks unchanged after my delta here, so we don't need to worry about it either.

Can anyone think of any other scenarios?

**Comment by DHowett-MSFT** [comment]
I can't think of any other reasonable scenarios, honestly. I'm comfortable with this.

It actually touches on something I wondered before: _can we use NUL to represent "unused buffer space"_ and translate it to space at the API boundary? It makes `MeasureRight` way easier for us if we can assume a NUL character is the right side of the line...

**Comment by DHowett-MSFT** [comment]
@msftbot make sure @miniksa signs off

**Comment by miniksa** [comment]
> I can't think of any other reasonable scenarios, honestly. I'm comfortable with this.
> 
> It actually touches on something I wondered before: _can we use NUL to represent "unused buffer space"_ and translate it to space at the API boundary? It makes `MeasureRight` way easier for us if we can assume a NUL character is the right side of the line...

This thought experiment is on my perpetual backlog for killing `MeasureRight`. I like the line of thought here, but I think I'd have to pursue it more deeply (by playing around with it) to fully understand the repercussions in a way that I can't manage in a simple PR comment.

**Comment by miniksa** [comment]
> Pains me to say, but should this be in WriteChars...?

Maybe in `WriteStream`. I'll reconsider it in November. This is fine for now.

**Comment by miniksa** [comment]
I'm fine simply because it's scoped to the VT engine. The only null-->space problems I can envision are around weird third party apps stuffing nulls into the storage space and retrieving it via `WriteConsoleOutput` and `ReadConsoleOutput`. Neither are involved in VT in any way (or at least they shouldn't be.)

**Pull request merged**